resolver/src/lookup_state.rs

```diff
@@ -239,61 +240,54 @@ impl<C: DnsHandle<Error = ResolveError> + 'static> QueryFuture<C> {
 
     fn handle_noerror(&mut self, mut message: Message) -> Poll<Records, ResolveError> {
         // seek out CNAMES
-        // TODO: figure out how to get rid of this clone
-        let mut cname_ttl = 0;
-        let mut was_cname = false;
-        let mut search_name: Name = self.query.name().clone();
-        while let Some(cname) = message.answers().iter().find(|r| {
-            r.rr_type() == RecordType::CNAME && r.name() == &search_name
-        })
-        {
-            was_cname = true;
-            cname_ttl = cname.ttl();
-            if let &RData::CNAME(ref name) = cname.rdata() {
-                if search_name == *name {
-                    break; // already searched for this name
-                } else {
-                    search_name = name.clone();
-                }
-            } else {
-                // now that is very odd...
-                warn!("Expected RData::CNAME in response record {:?}", cname);
-                break;
+        let (search_name, cname_ttl, was_cname) = {
+            let (search_name, cname_ttl, was_cname) = message.answers().iter()
+                .fold((Cow::Borrowed(self.query.name()), 0, false),
+                      |(search_name, cname_ttl, was_cname), r| {
+                          if let &RData::CNAME(ref cname) = r.rdata() {
+                              debug_assert_eq!(r.rr_type(), RecordType::CNAME);
+                              if search_name.as_ref() == r.name() {
+                                  return (Cow::Owned(cname.clone()), r.ttl(), true);
+                              }
+                          }
+                          (search_name, cname_ttl, was_cname)
+                      });
+
+            // After following all the CNAMES to the last one, try and lookup the final name
+            let records = message
+                .take_answers()
+                .into_iter()
+                .chain(message.take_additionals().into_iter())
+                .filter_map(|r| {
+                    let ttl = r.ttl();
+                    // TODO: disable name validation with ResolverOpts?
+                    // restrict to the RData type requested
+                    if self.query.query_type() == r.rr_type() && search_name.as_ref() == r.name() {
+                        Some((r.unwrap_rdata(), ttl))
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            if !records.is_empty() {
```

Comment:
can you throw a comment here that this is the result of a successfully found recordset?